questionnaire/questionnaire/utils/taskRunner/index.js

```diff
@@ -0,0 +1,139 @@
+'use strict';
+
+function getDataRef(ref, data) {
+    if (typeof ref === 'string' && ref[0] === '$') {
+        let objValue = data;
+
+        for (let i = 1, keys = ref.split('.'), len = keys.length; i < len; i += 1) {
+            objValue = objValue[keys[i]];
+        }
+
+        return objValue;
+    }
+
+    return ref;
+}
+
+async function wait(milliseconds) {
+    return new Promise(resolve => {
+        setTimeout(resolve, milliseconds);
+    });
+}
+
+function mutateObjectValues({key, value, valueTransformers, isMutableObject, ignoreValue}) {
+    if (ignoreValue && ignoreValue(value) === true) {
+        return value;
+    }
+
+    if (isMutableObject(value)) {
+        Object.entries(value).forEach(([k, v]) => {
+            value[k] = mutateObjectValues({
+                key: k,
+                value: v,
+                valueTransformers,
+                isMutableObject,
+                ignoreValue
+            });
+        });
+
+        return value;
+    }
+
+    const mutatedValue = valueTransformers.reduce((acc, transformValue) => {
+        return transformValue(key, acc);
+    }, value);
+
+    return mutatedValue;
+}
+
+function createTaskRunner({taskImplementations = {}, context = {}}) {
+    function hasTask(taskId) {
+        return Object.hasOwn(taskImplementations, taskId);
+    }
+
+    function isMutableObject(value) {
+        return value && typeof value === 'object' && hasTask(value.type) === false;
+    }
+
+    function ignoreValue(value) {
+        return value && typeof value === 'object' && hasTask(value.type) === true;
+    }
+
+    async function run(
+        factoryDefinition,
+        {
+            retries = factoryDefinition.retries === undefined ? 2 : factoryDefinition.retries,
+            retryCount = 0
+        } = {}
+    ) {
+        const taskState = {
+            id: factoryDefinition.id,
+            started: new Date().toISOString(),
+            ended: null,
+            result: null
+        };
```

Comment:
Should `taskState` have a `pending` state for tasks that are yet to be ran? This would make output explicit for the list of tasks in series. Thought of this whilst reading [this ](https://github.com/CriminalInjuriesCompensationAuthority/data-capture-service/pull/369/files#diff-de26e176c2c5d1b86ead521b94a322355bfcdd8b2725067599d3dabc48adc9f4R113)test.